src/main/kotlin/org/rust/ide/refactoring/move/RsMoveTopLevelItemsHandler.kt

```diff
@@ -84,3 +85,28 @@ class RsMoveTopLevelItemsHandler : MoveHandlerDelegate() {
         }
     }
 }
+
+private fun collectRelatedImplItems(containingMod: RsMod, items: List<RsItemElement>): List<RsImplItem> {
+    // For struct `Foo` we should collect:
+    // * impl Foo { ... }
+    // * impl ... for Foo { ... }
+    // * Maybe also `impl From<Foo> for Bar { ... }`?
+    //   if `Bar` belongs to same crate (but to different module from `Foo`)
+    //
+    // For trait `Foo` we should collect:
+    // * impl Foo for ... { ... }
+    return groupImplsByStructOrTrait(containingMod, items).values.flatten()
+}
+
+private fun groupImplsByStructOrTrait(containingMod: RsMod, items: List<RsItemElement>): Map<RsItemElement, List<RsImplItem>> {
+    return containingMod
+        .childrenOfType<RsImplItem>()
+        .mapNotNull { impl ->
+            val struct = (impl.typeReference?.type as? TyAdt)?.item
+            val trait = impl.traitRef?.path?.reference?.resolve() as? RsTraitItem
+            val relatedItem = struct?.takeIf { items.contains(it) } as RsItemElement?
+                ?: trait?.takeIf { items.contains(it) }
+            if (relatedItem != null) relatedItem to impl else null
```

Comment:
[minor] Probably more clear version
```suggestion
                ?: return@mapNotNull null
            relatedItem to impl
```